### PR TITLE
External display support on iOS

### PR DESCRIPTION
**Pull request by imxieyi**
It adds support for external display via Lightning-to-HDMI, AirPlay and Thunderbolt 3 on iOS. A proof-of-concept is shown as below:

iOS External Display Demo

**Task list:**
- [x] Automatically switch display when connected and disconnected
- [x] ~Automatically redirect audio output to external display if available~(This is automatic)
- ~Always show virtual controller on internal display~

### CODE REVIEW

File: ios/ViewController.mm [hunk 1]
```diff
@@ -144,7 +145,8 @@ - (void)subtleVolume:(SubtleVolume *)volumeView didChange:(CGFloat)value {
 - (void)viewDidLoad {
 	[super viewDidLoad];
 
-	self.view.frame = [[UIScreen mainScreen] bounds];
+    UIScreen* screen = [(AppDelegate*)[UIApplication sharedApplication].delegate screen];
```
Comment (unknownbrackets) [inline ios/ViewController.mm]: It'd be good to keep consistent with tabs throughout.  In PPSSPP, we use tabs so we just want to keep it consistent.

-[Unknown]

File: ios/ViewController.mm [hunk 2]
```diff
@@ -176,8 +179,8 @@ - (void)viewDidLoad {
 	}
 
 	g_dpi = (IS_IPAD() ? 200.0f : 150.0f) * scale;
-	g_dpi_scale_x = 240.0f / g_dpi;
-	g_dpi_scale_y = 240.0f / g_dpi;
+    g_dpi_scale_x = 1;//240.0f / g_dpi;
+    g_dpi_scale_y = 1;//240.0f / g_dpi;
```
Comment (unknownbrackets) [inline ios/ViewController.mm]: If possible, it'd be better to calculate `g_dpi` correctly and account for it in some way.  For example, if you connect this to a 6K display, the text may be unreadably small with scale of `1.0f`.

-[Unknown]

File: ios/ViewController.mm (hunk 1 above)
Comment (imxieyi) [reply ios/ViewController.mm]: Xcode use spaces for indentation by default. I have changed it to tabs.

File: ios/ViewController.mm (hunk 2 above)
Comment (imxieyi) [reply ios/ViewController.mm]: It seems that iOS does not provide API to get physical size of any display (both internal and external). I implemented scaling based on diagonal length instead.

**Comment by imxieyi** [comment]
@unknownbrackets As for the virtual controller, is it possible to render `TouchControlLayoutScreen` on separate `GraphicsContext`?

**Comment by unknownbrackets** [comment]
Hm.  In theory, that should be possible.  I'm not sure how it would work on iOS - I guess it would be a separate GLKView?

It sounds like (not sure) that would not really be a "separate" GL context, but really just a separate "backbuffer" we're drawing to.  Currently we draw everything to the surface bound via `[(GLKView*)self.view bindDrawable]` as if it was the backbuffer.

I think the cleanest way to implement this would be:

 * Add a `draw->GetNativeObject` flag to obtain a controller `Framebuffer *` uintptr_t.  This would be portable to other devices/implementations with two screens (i.e. Wii U, whatever.)
 * If the returned pointer `!= 0` (meaning there is a separate controller surface), we'd need to `draw->BindFramebufferAsRenderTarget(` it when drawing the onscreen controls (with updated viewport/etc.)
 * In `GLQueueRunner::PerformBindFramebufferAsRenderTarget`, we'd then detect this special `GLRFramebuffer *` and call something to bind the controller GLKView drawable.
 * We'd need to make sure the NativeTouch events are scaled by the controller surface size, not the display.
 * Easiest first step would be to move all the UI in `EmuScreen::renderUI()` to render to the controller surface instead of `nullptr` (the backbuffer.)
 * To ultimately split it up (on screen messages, save state previews, etc.), we'd need to detect the difference and create two separate root views in `EmuScreen::CreateViews()` (but only when the returned controller surface != 0.)
 * We'd also need to make sure to recreate views in this case when the display is connected/disconnected, which might already happen.  If it doesn't, would need `NativeMessageReceived("config_loaded", "");` or similar.

-[Unknown]

**Comment by imxieyi** [comment]
Rendering gamepad on separate screen is too complicated for now since PPSSPP has never considered this. Currently external display already works smoothly with physical keyboards and game controllers, which is well enough for this PR. Separate virtual gamepad can be implemented in a separate PR in the future.

**Comment by unknownbrackets** [comment]
This seems reasonable to me.  I don't have an iOS device to test on, but it seems cool and I'm in favor of merging.

I wonder if we can do something similar on Android.  That'd give more reasons to render the gamepad separately, and if we consider how it might work on Android we might come up with a cleaner implementation for iOS too.

-[Unknown]

**Review by unknownbrackets — APPROVED** [review]

File: CMakeLists.txt [hunk 3]
```diff
@@ -1988,6 +1988,7 @@ if(TargetBin)
 		endif()
 
 		if(IOS)
+            list(APPEND NativeAssets ios/Settings.bundle)
```
[the remark this answers is not in the document]
Comment (imxieyi) [reply CMakeLists.txt]: fixed

**Comment by dreampiggy** [comment]
I’ve installed the AppStore version of PPSSPP, and run on M1 iPad with 4K external display, seems it does not support Stage Manager’s “Full Screen” mode and always render as iPad native 4:3 resolution. Any idea about this ?

**Comment by hrydgard** [comment]
That probably needs some additional work. It may happen one day.

**Pull request merged**